### ISSUE TITLE
Path finding with edge attributes? 

### ISSUE THREAD

**Issue by FalafelGood** [JuliaGraphs/MetaGraphsNext.jl#14]
Hi, great project! Asking my question here since it looks like MetaGraphs.jl is on its way out.

I'm interested to know if shortest path routing can be done with respect to edge attributes?  For example suppose I have an attribute :active and I'm only interested in routing along edges where :active == true. What would be the best way for going about this in MetaGraphs or MetaGraphsNext? Does this feature seem worthwhile enough for me to make a PR?

Two reasons why I think this would be good:

1. Edge removal in adjacency lists is O(N) -- Small compared to something like a_star, but not constant either.
2. Removing edges for a single shortest path and adding them back in afterwards seems ugly and wasteful, particularly if the edges have lots of metadata.

Implementation ideas?

- Remap the distmx so any edge not conforming to a user's filter is assigned infinite weight.

Worth mentioning that I've never contributed to open source before. Your patience with me is appreciated! Thank you.

**Comment by bramtayl** [comment]
Hi Falafel,

Welcome! I'm not sure what the plans are for MetaGraphs but. That's a clever implementation idea. I think the most general way to do that would be to write a weight function which returns infinity under certain circumstances. We could make the struct mutable so that you could revise the weight function, but that might come at a performance cost?

**Comment by FalafelGood** [comment]
Oh wow, the MetaGraph struct in MetaGraphsNext is immutable? I've only been using MetaGraphs.jl up until now so I had no idea. That's fascinating. So vertices and edges are added by making another instance of the graph? It's certainly not my intention to compromise the performance of the package just for something I'd find useful. Probably this is a better idea for MetaGraphs.jl. What do you think?

**Comment by bramtayl** [comment]
Oh no, sorry if that's confusing. The references are immutable, but its content isn't. So you can add and delete as many vertices and edges you want without copying. It's just that you can't replace the graph with another graph entirely.

If you are interested in changing the weight function, we could add a `reuse` function which could allow you to construct a new MetaGraph with the same contents (copy by reference), but selectively replace some fields. Would that help?

**Comment by FalafelGood** [comment]
Ah of course! That makes a lot more sense, thanks for clarifying @bramtayl. Indeed, I think a reuse function to replace fields would be helpful. How should I get started?

**Comment by bramtayl** [comment]
If you want, you could make a pull request and I could comment. Or, I could make a pull request and you could comment. You can probably find instructions for making a pull request online.

**Comment by FalafelGood** [comment]
It would almost certainly be faster for you to make a PR and for me to comment since I'm not that clever at modifying other people's source code yet, but I want to get my hands dirty and learn by doing. (Plus, I don't want to take up your time) What I propose is that I have a stab at implementing that `reuse` function today. If you don't see a PR from me in the next 12-24 hours, you can assume that I haven't been able to do it, and then go ahead on your own. How does this sound?

**Comment by bramtayl** [comment]
Hi there, so guess you ran into trouble?

**Comment by FalafelGood** [comment]
Hi, sorry for the late reply. Yeah I tried something out about a week ago but put it down. Mea culpa. I'll set aside an hour today and pick it back up if it's not already done.

**Comment by gdalle** [comment]
@FalafelGood just to let you know I happen to need this exact functionality (shortest paths with forbidden edges) so I might have a go at it myself. However I think I will do it by reimplementing Dijkstra using an `active` attribute on the edge itself, since modifying the weights every time looks like a lot of work. I'll let you know how it goes

**Comment by cjprybol** [comment]
Stumbled upon this issue while learning my way through the differences between this package and MetaGraphs. Thanks for all of the development here!

Probably less efficient from a computational point of view but possibly quicker to implement, my thoughts on how to solve this were:
1. filter nodes and/or edges by metadata field of interest
2. call `induced_subgraph` to create a new graph with only the nodes/edges of interest
3. run shortest path algorithms on the induced subgraph

Is this a valid method? Is it going to be inefficient to the point of being impractical?

**Comment by gdalle** [comment]
@cjprybol it all depends how many times you need to run the shortest path routine. If it's only once, you'll do just fine.
What I ended up doing was recoding `a_star` with a `forbidden_vertices` argument. You can find an example [here](https://github.com/gdalle/MultiAgentPathFinding.jl/blob/giom/src/astar.jl), but I'm not sure it belongs in the MetaGraphsNext.jl library per se

**Comment by gdalle** [comment]
Closing since we probably won't work on it, if anything it might belong in Graphs.jl
